blob/src/main/java/com/bazaarvoice/emodb/blob/db/s3/S3RateLimiter.java

```diff
@@ -0,0 +1,160 @@
+package com.bazaarvoice.emodb.blob.db.s3;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.AmazonS3Exception;
+import com.bazaarvoice.emodb.blob.db.s3.config.S3ClientConfiguration;
+import com.google.common.base.Throwables;
+import com.google.common.reflect.AbstractInvocationHandler;
+import com.google.common.reflect.Reflection;
+import com.google.common.util.concurrent.RateLimiter;
+import org.apache.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+public class S3RateLimiter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(S3RateLimiter.class);
+
+    private final int maxAttempts;
+    private final SharedRateLimiter _l;
+
+    public S3RateLimiter(Clock clock, S3ClientConfiguration.RateLimitConfiguration rateLimitConfiguration) {
```

Comment:
Have we seen any rate limits from s3 during testing?

Reply:
nope